[1단계 - 자동차 경주 구현] 이오(이지우) 미션 제출합니다.

LJW25 opened the pull request:
안녕하세요, 우테코 5기 이오입니다.

코드 리뷰 감사드립니다!
첫 미션이라, 미숙한 점이 많을 것 같네요 :)

혹시 여유가 되신다면, 아래 질문을 읽어주시면 감사하겠습니다.

- 올바른 입력인지 확인하는 예외 처리를 입력인 InputView가 수행해야 할 일이라고 생각하여 InputView 내에서 구현하였습니다. 그런데 생각하기에 따라 이 또한 입력값을 판단하는 로직이기 때문에 별도의 패키지로 구분해야하나? 라는 고민을 계속하고 있습니다. View에서 예외에 대한 검증까지 포함하여도 괜찮을까요?
- 단위 테스트를 구현 할 때, private은 메소드를 직접 호출할 수 없어 테스트 하기가 어렵기 때문에 InputViewTest에서 각 검증 메소드를 테스트하는 것이 아닌 입력 메소드를 테스트하는 방식으로 구현하였습니다. 해당 방식이 적절할까요? 혹은 검증 메소드 각각을 테스트하기 위해 다른 방식을 사용해야 할까요?

sh111-coder committed:
test(StringStudyTest) : String 분리 테스트

sh111-coder committed:
test(StringStudyTest): String 특정 문자 제거 테스트

sh111-coder committed:
test(StringStudyTest) : 특정 문자 위치 찾기 문자열로 특정 문자 찾기 테스트

sh111-coder committed:
test(StringStudyTest) : 문자열로 특정 문자 찾기 테스트

sh111-coder committed:
test(StringStudyTest) : 인덱스로 특정 문자 찾기 예외 테스트

sh111-coder committed:
test(SetStudyTest) : Set 크기 확인 테스트

sh111-coder committed:
test(SetStudyTest) : Set 값 존재 확인 테스트

sh111-coder committed:
test(SetStydtTest) : Set 입력 값에 따라 결과 값이 다른 경우 테스트

sh111-coder committed:
docs(README.md) : 기능 목록 초안 작성

sh111-coder committed:
feat(InputView) : InputView 클래스, 메서드 선언

sh111-coder committed:
feat(OutputView) : OutputView 클래스, 메소드 선언

sh111-coder committed:
feat(Car) : Car 클래스, 메소드 선언

sh111-coder committed:
feat(RacingCars) : RacingCars 클래스, 메서드 선언

sh111-coder committed:
feat(RandomNumberGenerator) : RandomNumberGenerator 클래스, 메소드 선언

sh111-coder committed:
feat(OutputView) : 자동차 이름 입력 문구 출력 기능 구현

sh111-coder committed:
feat(OutputView) : 시도할 횟수 입력 문구 출력 기능 구현

sh111-coder committed:
feat(OutputView) : 실행 결과 문구 출력 기능 구현

sh111-coder committed:
feat(OutputView) : 자동차들의 현재 위치 출력 기능 구현

sh111-coder committed:
feat(OutputView) : 최종 우승자 출력 기능 구현

sh111-coder committed:
feat(InputView) : 자동차 이름이 빈 값인 경우 예외 처리 구현

sh111-coder committed:
feat(InputView) : 자동차 이름이 중복되는 경우 예외 처리 구현

sh111-coder committed:
feat(InputView) : 구분한 각 자동차 이름이 5글자 초과인 경우 예외 처리 구현

sh111-coder committed:
test(InputViewTest) : 자동차 이름 입력 기능 테스트

sh111-coder committed:
feat(InputView) : 자동차 이름 입력 기능 구현

sh111-coder committed:
feat(InputView) : 시도할 횟수가 빈 값인 경우 예외 처리 구현

sh111-coder committed:
feat(InputView) : 시도할 횟수가 정수가 아니면 예외 처리 기능 구현

sh111-coder committed:
feat(InputView) : 시도할 횟수 입력 기능 구현

sh111-coder committed:
test(InputViewTest) : 시도할 횟수 입력 테스트

sh111-coder committed:
rename : racingcar 패키지 생성, 이동

sh111-coder committed:
feat(RandomNumberGenerator) : 0~9 사이 랜덤 값 생성 기능 구현

sh111-coder committed:
feat(Car) : 자동차 전진 여부 결정 기능 구현

sh111-coder committed:
feat(Car) : 자동차 전진 기능 구현

sh111-coder committed:
rename(InputViewTest) : InputViewTest 패키지 이동

sh111-coder committed:
test(CarTest) : Car 메소드 테스트

sh111-coder committed:
test(RacingCarsTest) : 우승자 선정 기능 테스트

sh111-coder committed:
feat(RacingCars) : 우승자 선정 기능 구현

sh111-coder committed:
fix(OutputView) : 우승자 출력 시 배열로 표시되는 오류 fix

sh111-coder committed:
test(RacingCarsTest) : 축약 refactoring

sh111-coder committed:
feat(RacingCarController) : RacingCarController 구현

sh111-coder committed:
feat(Application) : Application 구현

LJW25 committed:
refactor(InputView) : 상수 순서 변경

LJW25 committed:
test : 사용하지 않는 패키지 제거

joseph415 commented on src/main/java/racingcar/controller/RacingCarController.java:
```diff
@@ -0,0 +1,78 @@
+
+public class RacingCarController {
+
+    OutputView outputView = OutputView.getInstance();
```
View 객체라 private 따로 안달아 주신것 같은데 맞을까요?~

LJW25 replied:
아래에 답한대로, 놓치고 있던 부분이라 private 추가하였습니다!

joseph415 commented on src/test/java/racingcar/model/CarTest.java:
```diff
@@ -0,0 +1,68 @@
+package racingcar.model;
```
테스트 작성 꼼꼼하게 해주셨네요👍

joseph415 commented on src/main/java/racingcar/controller/RacingCarController.java:
```diff
@@ -0,0 +1,78 @@
+
+    OutputView outputView = OutputView.getInstance();
+    InputView inputView = InputView.getInstance();
+    RacingCars racingCars;
```
접근 제어자 추가하면 좋을 것 같습니다~

LJW25 replied:
해당 부분 놓치고 있었네요! private로 추가하였습니다.

joseph415 commented:
> View에서 예외에 대한 검증까지 포함하여도 괜찮을까요?

InputView 내에서 해도 좋고, view 패키지 내부에 validate 패키지로 따로 빼서 뷰전용 검증 객체를 만들어도 좋아고 생각합니다! 책임분리를 연습할겸 분리해보는 것도 좋을 것 같아요!

> 해당 방식이 적절할까요? 혹은 검증 메소드 각각을 테스트하기 위해 다른 방식을 사용해야 할까요?

private method 이기 때문에 외부에는 readCarNames만 노출됩니다. 사용자 입장에서 readCarNames의 역할 검증이 필요할것이기 때문에, 현 방식대로 하는게 맞다고 생각합나다.

검증메서드를 따로 두고 관리하면서 테스트할 필요가 있다고 생각되면, 첫번째 답변처럼 validator를 따로 두고 객체분리를 하고 테스트를 작성하면 될 것 같아요

joseph415 commented on src/main/java/racingcar/controller/RacingCarController.java:
```diff
@@ -0,0 +1,78 @@
+        outputView.printRacingResultMessage();
+        for (int repeatIndex = 0; repeatIndex < tryNum; repeatIndex++) {
+            List<Car> currentCars = racingCars.getCars();
+            for (Car currentCar : currentCars) {
```
들여쓰기는 한단계만 하도록 구현해보는것 어떤가요~?

LJW25 replied:
한 회 마다 수행되는 역할인 두번째 for문을 별도의 메소드로 분리하였습니다! :)

joseph415 commented on src/main/java/racingcar/model/Car.java:
```diff
@@ -0,0 +1,49 @@
+        if (this.position == maxPositionCar.getPosition()) {
+            return this;
+        }
+        return null;
```
samePositionCar 를 get 하는 메서드인데 null 이 리턴되는 상황을 내부확인하지 않으면 외부에서 사용자 입장에서는 모를것 같은데요. 네이밍을 변경해보는건 어떤가요?

요건 개인적인 의견이라ㅎㅎ 이오가 맞다고 생각하시면, 그대로 진행해주셔도 좋아요~

LJW25 replied:
해주신 조언에 더해, 페어가 받은 리뷰인 null 사용을 자제하라는 조언을 함께 고려하여 해당 메소드의 반환 형식을 boolean으로 수정하고 메소드의 이름을 isSamePostionCar로 변경하였습니다!

joseph415 commented on src/main/java/racingcar/model/RacingCars.java:
```diff
@@ -0,0 +1,42 @@
+        List<Car> winners = new ArrayList<>();
+        for (Car car : cars) {
+            Car sameMaxPositionCar = car.getSamePositionCar(maxPositionCar);
+            if (sameMaxPositionCar != null) {
```
들여쓰기 한단계만 해보도록 수정해보는거 어떤가요?~

LJW25 replied:
for문과 if문을 사용하는 대신 stream.filter()를 사용하여 간결하게 표현될 수 있도록 수정하였습니다 :)

joseph415 reviewed (commented):
안녕하세요. 리뷰어 서브웨이입니다. 미션 구현 너무 잘해주셨네요! 💯

간단한 코멘트 남겼습니다~

LJW25 committed:
refactor(RacingCarController) : 접근 제어자 추가 및 선언 순서 변경

LJW25 committed:
refactor(RacingCarController) : race 메소드 분리

LJW25 committed:
refactor(Car) : getSamePositionCar 반환 타입 및 메소드 이름 변경

- 기존의 Car 를 반환하던 형식에서 같은 위치인지 여부(boolean)를 반환하는 형식으로 변경
- 변경된 형식에 따라 메소드 이름을 getSamePositionCar에서 isSamePositionCar로 변경

LJW25 committed:
refactor(RacingCars) : getMaxPositionCars 메소드 간결화

LJW25 committed:
test(CarTest) : 수정된 메소드 반영

LJW25 commented:
> 검증메서드를 따로 두고 관리하면서 테스트할 필요가 있다고 생각되면, 첫번째 답변처럼 validator를 따로 두고 객체분리를 하고 테스트를 작성하면 될 것 같아요

답변 감사드립니다! 다음번엔 vaildator 객체를 분리해서 사용하는 연습도 해 보고, 더 나은 방법을 찾아봐야겠네요 :)

joseph415 commented on src/main/java/racingcar/RandomNumberGenerator.java:
```diff
@@ -0,0 +1,13 @@
+
+import java.util.Random;
+
+public class RandomNumberGenerator {
```
RandomNumberGenerator가 단순히 utils로서 랜덤난수 생성기로 사용될 수 도 있지만, 만약 자동차 전진 규칙이 랜덤이 아니고 다른 규칙으로 바뀌게 되면 어떻게 될까요?

추가적으로, 해당 규칙은 자동차가 전진하는 중요한 비지니스 로직이 될 수도 있을텐데요. utils 이 아닌 도메인 객체로서 Generator를 interface 로 만들어 보는것도 객체지향적인 코드를 만드는데 도움이 될 수 있을 것 같아서 의견을 남겨요ㅎ

그렇게 된다면, Generator에 대한 테스트도 유연하게 할 수 있을 거에요!

joseph415 commented on src/main/java/racingcar/controller/RacingCarController.java:
```diff
@@ -0,0 +1,82 @@
+        outputView.printWinners(winners);
+    }
+
+    private void race(int tryNum) {
```
private 함수들의 순서가 뒤섞여 있는것 같은데, 혹시 순서를 고려하지 않으셨다면, 내려가기 규칙에 맞추면, 좀 더 읽기 쉬워질 것 같아요!

[내려가기 규칙 참고](https://hyojaedev.tistory.com/40)

LJW25 replied:
메소드를 분리하는 과정에서 순서를 고려하지 않고 선언한 실수가 있었네요! 알려주신 사이트 참고하여, 내려가기 규칙으로 수정하였습니다. 감사합니다!

joseph415 commented on src/main/java/racingcar/controller/RacingCarController.java:
```diff
@@ -0,0 +1,82 @@
+        }
+    }
+
+    private static void tryOneTime(List<Car> currentCars) {
```
해당 메서드는 왜 static 함수로 되어있나요?

LJW25 replied:
같은 클래스 내의 race 메소드에서만 호출되기 때문에 불필요하게 사용된 static이었습니다. 삭제하였습니다!

joseph415 commented on src/main/java/racingcar/model/Car.java:
```diff
@@ -0,0 +1,49 @@
+    }
+
+    public boolean canMoving(int randomValue) {
+        if (randomValue >= MIN_MOVING_NUM) {
```
한줄로 줄일 수 있을 것 같네요!

LJW25 replied:
바로 반환하는 형식으로 수정하였습니다! isSamePositionCar 메소드도 비슷한 형식이어 동일하게 수정하였습니다 :)

joseph415 commented on src/main/java/racingcar/model/RacingCars.java:
```diff
@@ -0,0 +1,39 @@
+    }
+
+    public List<Car> getCars() {
+        return Collections.unmodifiableList(cars);
```
unmodifiableList 👍

joseph415 reviewed (approved):
안녕하세요 이오! 

리뷰 잘 반영해주셨는데요.💯 1단계는 마무리해도 될 것 같아서 merge 하겠습니다! 고생 하셨습니다!
추가적으로 간단한 코멘트 하나 남겼습니다.
